File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/example/localEstimator/TransferLearning.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intel.analytics.zoo.examples.localEstimator

import com.intel.analytics.bigdl.nn._
import com.intel.analytics.bigdl.optim.{Adam, Loss, Top1Accuracy}
import com.intel.analytics.zoo.pipeline.api.net.TFNet
import com.intel.analytics.zoo.pipeline.estimator.LocalEstimator
import org.slf4j.LoggerFactory
import scopt.OptionParser

case class TransferLearningParams(imageDirPath: String = "./",
                                  modelPath: String = "./",
                                  inputs: String = "",
                                  outputs: String = "",
                                  batchSize: Int = 132,
                                  epoch: Int = 20,
                                  threadNum: Int = 10)

object TransferLearning {

  val logger = LoggerFactory.getLogger(getClass)

  def main(args: Array[String]): Unit = {
    val parser = new OptionParser[TransferLearningParams]("TransferLearning Example") {
      opt[String]('d', "imageDirPath")
        .required()
        .text("The directory of cifar10 dataset")
        .action((x, c) => c.copy(imageDirPath = x))
      opt[String]('m', "modelPath")
        .required()
        .text("The path of model")
        .action((x, c) => c.copy(modelPath = x))
      opt[String]('i', "inputs")
        .required()
        .text("The inputs of model")
        .action((x, c) => c.copy(inputs = x))
      opt[String]('o', "outputs")
        .required()
        .text("The outputs of model")
        .action((x, c) => c.copy(outputs = x))
      opt[Int]('b', "batchSize")
        .required()
        .text("The number of batchSize")
        .action((x, c) => c.copy(batchSize = x))
      opt[Int]('e', "epoch")
        .required()
        .text("The number of epoch")
        .action((x, c) => c.copy(epoch = x))
      opt[Int]('t', "threadNum")
        .required()
        .text("The number of threadNum")
        .action((x, c) => c.copy(threadNum = x))
    }

    parser.parse(args, TransferLearningParams()).map { params =>
      logger.info(s"params parsed as $params")

      val modelPath = params.modelPath
      val inputs = params.inputs.split(",")
      val outputs = params.outputs.split(",")
      val imageDirPath = params.imageDirPath
      val batchSize = params.batchSize
      val epoch = params.epoch
      val threadNum = params.threadNum

      val originalModel = TFNet.fromSavedModel(modelPath, inputs, outputs)
      println(originalModel)

      val model = Sequential[Float]()
      model.add(Transpose[Float](Array((2, 4), (2, 3))))
      model.add(Contiguous[Float]())
      model.add(originalModel)
      model.add(Transpose[Float](Array((2, 4), (3, 4))))
      model.add(Contiguous[Float]())
      model.add(new SpatialAveragePooling[Float](2, 2, globalPooling = true))
      model.add(new View[Float](2048).setNumInputDims(3))
      model.add(new Linear[Float](2048, 2))

      println(model)

      val criterion = new CrossEntropyCriterion[Float]()
      val adam = new Adam[Float]()
      val validations = Array(new Top1Accuracy[Float], new Loss[Float])
      val localEstimator = LocalEstimator(model, criterion, adam, validations, threadNum)
      println(s"LocalEstimator loaded as $localEstimator")

      // use only a little data for transfer learning
      val trainData = Cifar10DataLoader.loadTrainData(imageDirPath)
        .filter(_.label() <= 2).slice(0, 10 * batchSize)
      val testData = Cifar10DataLoader.loadTestData(imageDirPath)
        .filter(_.label() <= 2).slice(0, 10 * batchSize)

      localEstimator.fit(trainData,
        testData,
        ImageProcessing.labeledBGRImageToMiniBatchTransformer,
        batchSize,
        epoch)
    }
  }

}

File: python/orca/example/learn/tf/transfer_learning/transfer_learning.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
#
# Copyright 2019 The TensorFlow Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# This example shows how to classify cats vs dogs images
# by using transfer learning from a pre-trained network on Orca,
# and is adapted from
# https://github.com/tensorflow/docs/blob/master/site/en/r1/tutorials/images/transfer_learning.ipynb

import os
from os.path import exists
from os import makedirs
import argparse

import tensorflow as tf
import tensorflow_datasets as tfds
from tensorflow import keras

from bigdl.orca import init_orca_context, stop_orca_context
from bigdl.orca.learn.tf.estimator import Estimator

#import os
#for filepath,dirnames,filenames in os.walk('/data/datasets/cats_and_dogs_filtered'):
#    for filename in filenames:
#        print(os.path.join(filepath,filename))

parser = argparse.ArgumentParser()
parser.add_argument('--cluster_mode', type=str, default="local",
                    help='The mode for the Spark cluster. local, yarn or spark-submit.')
parser.add_argument('--data_dir', type=str, default="./dataset", help='The path of datesets.')
parser.add_argument('--batch_size', type=int, default=64, help='The training batch size')
parser.add_argument('--epochs', type=int, default=2, help='The number of epochs to train for')
parser.add_argument('--download_url', type=str, default="https://storage.googleapis.com/mledu-datasets/cats_and_dogs_filtered.zip",
                    help="The url of cats_and_dogs_filtered.zip.")
args = parser.parse_args()
cluster_mode = args.cluster_mode

dataset_dir = args.data_dir
download_url = args.download_url
if not exists(dataset_dir):
    makedirs(dataset_dir)
zip_file = tf.keras.utils.get_file(
    origin=download_url,
    fname="cats_and_dogs_filtered.zip", extract=True, cache_dir=dataset_dir)
base_dir, _ = os.path.splitext(zip_file)

if cluster_mode == "local":
    init_orca_context(cluster_mode="local", cores=4, memory="3g")
elif cluster_mode.startswith("yarn"):
  if cluster_mode == "yarn-client":
    additional = dataset_dir
    init_orca_context(cluster_mode="yarn-client", num_nodes=2, cores=2, driver_memory="3g",
                      additional_archive=additional)
  else:
    additional = dataset_dir
    init_orca_context(cluster_mode="yarn-cluster", num_nodes=2, cores=2, driver_memory="3g",
                      additional_archive=additional)
elif cluster_mode == "spark-submit":
    init_orca_context(cluster_mode="spark-submit")                      
else:
    print("init_orca_context failed. cluster_mode should be one of 'local', 'yarn' and 'spark-submit' but got "
          + cluster_mode)

train_dir = os.path.join(base_dir, 'train')
validation_dir = os.path.join(base_dir, 'validation')

# Directory with our training cat pictures
train_cats_dir = os.path.join(train_dir, 'cats')
print('Total training cat images:', len(os.listdir(train_cats_dir)))

# Directory with our training dog pictures
train_dogs_dir = os.path.join(train_dir, 'dogs')
print('Total training dog images:', len(os.listdir(train_dogs_dir)))

# Directory with our validation cat pictures
validation_cats_dir = os.path.join(validation_dir, 'cats')
print('Total validation cat images:', len(os.listdir(validation_cats_dir)))

# Directory with our validation dog pictures
validation_dogs_dir = os.path.join(validation_dir, 'dogs')
print('Total validation dog images:', len(os.listdir(validation_dogs_dir)))

image_size = 160  # All images will be resized to 160x160
batch_size = args.batch_size


# Rescale all images by 1./255 and apply image augmentation
def _parse_function(img, label):
    image_resized = tf.image.resize(img, [image_size, image_size])
    return tf.cast(image_resized, dtype=tf.float32) / 255., tf.cast(label, dtype=tf.uint8)


builder = tfds.ImageFolder(base_dir)
ds = builder.as_dataset(shuffle_files=True, as_supervised=True)
train_dataset = ds['train'].map(_parse_function)

validation_dataset = ds['validation'].map(_parse_function)

IMG_SHAPE = (image_size, image_size, 3)

# Create the base model from the pre-trained model MobileNet V2
base_model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
                                               include_top=False,
                                               weights='imagenet')

base_model.trainable = False
base_model.summary()

model = tf.keras.Sequential([
    base_model,
    keras.layers.GlobalAveragePooling2D(),
    keras.layers.Dense(1, activation='sigmoid')
])
model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=0.0001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.summary()
len(model.trainable_variables)

epochs = args.epochs
est = Estimator.from_keras(keras_model=model)
est.fit(train_dataset,
        batch_size=batch_size,
        epochs=epochs,
        validation_data=validation_dataset
        )
result = est.evaluate(validation_dataset)
print(result)

base_model.trainable = True
# Let's take a look to see how many layers are in the base model
print("Number of layers in the base model: ", len(base_model.layers))

# Fine tune from this layer onwards
fine_tune_at = 100

# Freeze all the layers before the `fine_tune_at` layer
for layer in base_model.layers[:fine_tune_at]:
    layer.trainable = False

model = tf.keras.Sequential([
    base_model,
    keras.layers.GlobalAveragePooling2D(),
    keras.layers.Dense(1, activation='sigmoid')
])
model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=2e-5),
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.summary()
print(len(model.trainable_variables))

est = Estimator.from_keras(keras_model=model)
est.fit(data=train_dataset,
        batch_size=batch_size,
        epochs=epochs,
        validation_data=validation_dataset
        )
print("==>unfreeze")
result = est.evaluate(validation_dataset)
print(result)
stop_orca_context()